Subject: [PATCH] Sometimes Camera2 APK hung on record stop

Blocked on AudioSource::waitOutstandingEncodingFrames_l().
In MediaCodecSource, when process kWhatStop event, if already met EOS,
should also call mPuller->stop() to return back media buffers.

diff --git a/media/libstagefright/MediaCodecSource.cpp b/media/libstagefright/MediaCodecSource.cpp
index d808e5b..589dbf2 100644
--- a/media/libstagefright/MediaCodecSource.cpp
+++ b/media/libstagefright/MediaCodecSource.cpp
@@ -999,6 +999,12 @@
             // if we already reached EOS, reply and return now
             ALOGI("encoder (%s) already stopped",
                     mIsVideo ? "video" : "audio");
+
+            // There maybe media buffers in the queue, should return back.
+            // Or the AudioSource will block on waitOutstandingEncodingFrames_l()
+            if (!(mFlags & FLAG_USE_SURFACE_INPUT))
+                mPuller->stop();
+
             (new AMessage)->postReply(replyID);
             break;
         }
